[BEAM-6674] Add schema support to JdbcIO read

charithe opened the pull request:
Adds support for directly reading `Row` objects from a JDBC datasource by automatically converting the database schema to a Beam schema.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)
Python | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/)<br>[![Build Status](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Python3_Verify/lastCompletedBuild/) | --- | [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) <br> [![Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_PVR_Flink_Cron/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [![Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/badge/icon)](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

charithe commented:
R: @reuvenlax 
R: @jbonofre

reuvenlax commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -188,6 +191,15 @@
+        .setFetchSize(DEFAULT_FETCH_SIZE)
+        .setOutputParallelization(true)
+        .build();
+  }
```
This shouldn't be limited to just reading Row objects, as any type can have a schema. So you should be able to have a readWithSchema(Class<T> clazz) and get back a PCollection<T> (if the SchemaRegistry contains a schema for clazz, and if that schema matches the one inferred at the source). readRows is still useful but can be a wrapper around this.

charithe replied:
Do you prefer an explicit `readWithSchema` method or do you prefer it to be implicit?. That is:

- Make the JDBC row mapper optional
- If the user provides a JDBC row mapper, then:
    - If the output type of the row mapper exists in the registry, attach that schema to the output PCollection
    - Otherwise, the output PCollection will not have a schema
- If the user does not provide a JDBC row mapper, then:
    - If the output type exists in the schema registry and is compatible with the inferred schema from the database, generate our own row mapper to map the resultset and attach the schema to the output PCollection
    - Otherwise, throw an exception

reuvenlax replied:
I think we can make it implicit in this case.

charithe replied:
I looked into this a little bit more and I think I may have misunderstood a few things.

I had the idea that if I get the `TypeDescriptor` of the PCollection, I could query the schema registry to get the necessary information to attach the schema. However, this doesn't seem to work in practice. Of course, I could be doing something wrong there but I came to the realisation that if it was indeed a viable solution, then it would have been already implemented at the core as it would provide a generic way to implicitly attach schemas to any PCollection that contains suitable element types.

Now I am not quite sure what benefit a `readWithSchema(clazz)` method would have. If the user does need schema support, they could just as well call `setSchema` on the PCollection themselves, can't they? We can reduce the boilerplate involved but that is not specific to the JDBC IO module and should probably go in to the core schema module as a utility function.

Since the user is responsible for defining how to map the ResultSet to a type of their choosing, they could potentially do some field transforms during the mapping and produce an output type whose schema doesn't match up with the schema inferred from the raw data. Therefore I feel like comparing the two schemas will not be very useful either.

What are your thoughts?

reuvenlax replied:
I was thinking that this would allow us to not have to create a separate fork of the Read PTransform. Just keep one Read<T>, which happens to support schemas (as long as T is in the SchemaRegistry); yes the user can call setSchema themselves, however forcing the user to do a lookup in the SchemaRegistry themselves in ugly, and we should make that implicit. Then readRows will just return Read<Row>, so we should not need a separate class.

Does that work?

charithe replied:
So I assumed that it would be easy to obtain the schema for `T` and implicitly attach it to the PCollection. However, this appears to be difficult to do reliably unless the user explicitly provides a reference to `Class<T>` as you correctly suggested. 

Given that the user has to make a conscious effort to enable schema support by providing a reference to `Class<T>`, wouldn't it make sense to introduce a generic utility function such as `PCollections.withSchema(PCollection<T>, Class<T>)` that would work for any kind of PCollection and not just the output of JDBC IO?

reuvenlax commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java:
```diff
@@ -391,6 +403,123 @@ DataSource buildDatasource() {
+  /** Implementation of {@link #readRows()}. */
+  @AutoValue
+  @Experimental(Experimental.Kind.SCHEMAS)
+  public abstract static class ReadRows extends PTransform<PBegin, PCollection<Row>> {
```
can we share the part that's common with Read? I'm worried that that people will end up adding features to one codepath and not the other.

charithe replied:
Do you have any suggestions on how to do this? I wasn't happy with the duplication either but I couldn't see a nice way to share state between AutoValue classes. Both the existing `Read` and `ReadAll` classes are pretty much identical to each other as well so I assumed that was a known issue.

reuvenlax commented:
I think we can make it implicit in this case.

On Sun, Jun 2, 2019 at 3:51 AM Charith Ellawala <notifications@github.com>
wrote:

> *@charithe* commented on this pull request.
> ------------------------------
>
> In sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java
> <https://github.com/apache/beam/pull/8725#discussion_r289636876>:
>
> > @@ -188,6 +191,15 @@
>          .build();
>    }
>
> +  /** Read Beam {@link Row}s from a JDBC data source. */
> +  @Experimental(Experimental.Kind.SCHEMAS)
> +  public static ReadRows readRows() {
> +    return new AutoValue_JdbcIO_ReadRows.Builder()
> +        .setFetchSize(DEFAULT_FETCH_SIZE)
> +        .setOutputParallelization(true)
> +        .build();
> +  }
>
> Do you prefer an explicit readWithSchema method or do you prefer it to be
> implicit?. That is:
>
>    - Make the JDBC row mapper optional
>    - If the user provides a JDBC row mapper, then:
>       - If the output type of the row mapper exists in the registry,
>       attach that schema to the output PCollection
>       - Otherwise, the output PCollection will not have a schema
>    - If the user does not provide a JDBC row mapper, then:
>       - If the output type exists in the schema registry and is
>       compatible with the inferred schema from the database, generate our own row
>       mapper to map the resultset and attach the schema to the output PCollection
>       - Otherwise, throw an exception
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/8725?email_source=notifications&email_token=AFAYJVIECRY3TJWPEAWLJTDPYOQ2BA5CNFSM4HRE7V22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB2KJYYI#discussion_r289636876>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFAYJVKRJFTURD4UNFJIYLTPYOQ2BANCNFSM4HRE7V2Q>
> .
>

charithe commented on sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/SchemaUtil.java:
```diff
@@ -0,0 +1,321 @@
+
+  /*
+   * Mapping of primitive JDBC types to Beam schema types.
+   * TODO: Add support for BLOB and CLOB types
```
Done: https://issues.apache.org/jira/browse/BEAM-7485

reuvenlax commented:
What you suggested already works via Schema inference. If you have
PCollection<T> and T has a registered schema, then Beam will infer it
(assuming that T is not erased for some reason). However to convert the
Jdbc class to T requires knowing the actual type in the IO.

On Fri, Jun 7, 2019 at 4:42 AM Charith Ellawala <notifications@github.com>
wrote:

> *@charithe* commented on this pull request.
> ------------------------------
>
> In sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java
> <https://github.com/apache/beam/pull/8725#discussion_r291556575>:
>
> > @@ -188,6 +191,15 @@
>          .build();
>    }
>
> +  /** Read Beam {@link Row}s from a JDBC data source. */
> +  @Experimental(Experimental.Kind.SCHEMAS)
> +  public static ReadRows readRows() {
> +    return new AutoValue_JdbcIO_ReadRows.Builder()
> +        .setFetchSize(DEFAULT_FETCH_SIZE)
> +        .setOutputParallelization(true)
> +        .build();
> +  }
>
> So I assumed that it would be easy to obtain the schema for T and
> implicitly attach it to the PCollection. However, this appears to be
> difficult to do reliably unless the user explicitly provides a reference to
> Class<T> as you correctly suggested.
>
> Given that the user has to make a conscious effort to enable schema
> support by providing a reference to Class<T>, wouldn't it make sense to
> introduce a generic utility function such as PCollections.withSchema(PCollection<T>,
> Class<T>) that would work for any kind of PCollection and not just the
> output of JDBC IO?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/apache/beam/pull/8725?email_source=notifications&email_token=AFAYJVLYKZV6GQMKO6KPPHLPZJCQTA5CNFSM4HRE7V22YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB242OVA#discussion_r291556575>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFAYJVKQ3KYP6SUPZL2TFU3PZJCQTANCNFSM4HRE7V2Q>
> .
>

charithe commented:
I modified the code to attach the schema to PCollection if the type is registered in the registry (https://github.com/apache/beam/pull/8725/commits/046acd0c7f29e90beffb2fcf9a631f3a6bdc5da1). Is that what you were expecting?

Regarding the `readRows` method I introduced which automatically generates a row mapper based on the inferred schema, I am struggling to figure out a nice way of exposing it without some duplication of the `Read` class. 
- Reusing the `Read` class by making the row mapper and coder optional would be confusing for users because they have to remember to explicitly create a `Read<Row>` instance and be aware of the things they shouldn't set in that instance.
- Extending the `Read` class won't work because the visibility of `withRowMapper` and `withCoder` methods cannot be downgraded.
- As only row mapper and coder are optional, there are still quite few other arguments that the user needs to pass. So making `readRows` accept them all in the argument list would be unfriendly as well.

I may be over-thinking this part. Do you have a suggestion on how you would design the API in this case?

reuvenlax commented:
inferBeamSchema should be called in the path where you lookup the schema registry as well I think. In that scenario you want to make sure that the schema is compatible.

charithe commented:
If the user provides their own row mapper to convert the resultset to some type `T`, then I don't see why the schema of `T` should be compatible with the inferred schema of the resultset. For example, the query might have separate fields to hold `first_name` and `last_name` but during the mapping the user might choose to combine those two fields to create a single field named `full_name`.

reuvenlax commented:
Good point, I agree.

charithe commented:
Thanks Reuven. I think the only other point remaining is the duplication of code between `Read` and `ReadRows`. As I mentioned in a comment above, I am struggling to find a good solution for that problem. Do you have any suggestions on how to address that problem?

reuvenlax commented:
I will think about the code duplication a bit more, but I don't think it needs to be resolved for this PR. Could you file a JIRA ticket for that, and I will merge this PR.

reuvenlax commented:
lgtm

terekete commented:
Hey all -  I was wondering if there is an example or docs on how to use the the new implementation to read rows using the Row Coder and Row Mapper and auto-infer the beam schema ?
